core/arch/x86/clean_call_opt.c

```diff
@@ -113,32 +128,38 @@ analyze_callee_regs_usage(dcontext_t *dcontext, callee_info_t *ci)
 
     /* check if callee read aflags from caller */
     /* set it false for the case of empty callee. */
-    ci->read_flags = false;
-    for (instr = instrlist_first(ilist); instr != NULL; instr = instr_get_next(instr)) {
-        uint flags = instr_get_arith_flags(instr, DR_QUERY_DEFAULT);
-        if (TESTANY(EFLAGS_READ_6, flags)) {
-            ci->read_flags = true;
-            break;
+    if (ZMM_ENABLED()) {
+        LOG(THREAD, LOG_CLEANCALL, 2,
+            "CLEANCALL: AVX-512 enabled, forced to save aflags\n", ci->start);
+    } else {
+        ci->read_flags = false;
+        for (instr = instrlist_first(ilist); instr != NULL;
+             instr = instr_get_next(instr)) {
+            uint flags = instr_get_arith_flags(instr, DR_QUERY_DEFAULT);
+            if (TESTANY(EFLAGS_READ_6, flags)) {
+                ci->read_flags = true;
+                break;
+            }
+            if (TESTALL(EFLAGS_WRITE_6, flags))
+                break;
+            if (instr_is_return(instr))
+                break;
+            if (instr_is_cti(instr)) {
+                ci->read_flags = true;
+                break;
+            }
         }
-        if (TESTALL(EFLAGS_WRITE_6, flags))
-            break;
-        if (instr_is_return(instr))
-            break;
-        if (instr_is_cti(instr)) {
-            ci->read_flags = true;
-            break;
+        if (ci->read_flags) {
+            LOG(THREAD, LOG_CLEANCALL, 2,
+                "CLEANCALL: callee " PFX " reads aflags from caller\n", ci->start);
         }
     }
-    if (ci->read_flags) {
-        LOG(THREAD, LOG_CLEANCALL, 2,
-            "CLEANCALL: callee " PFX " reads aflags from caller\n", ci->start);
-    }
 
     /* If we read or write aflags, we need to reserve a slot to save them.
      * We may or may not use the slot at the call site, but it needs to be
      * reserved just in case.
      */
-    if (ci->read_flags || ci->write_flags) {
+    if (ci->read_flags || ci->write_flags | ZMM_ENABLED()) {
```

Comment:
Just set ci->read flags and avoid all these changes?

Reply:
Hmm this sounds dangerous to me. I tried to leave as-is, just adding ZMM_ENABLED(). It's not that much, just a few lines. Sure? LMK